Access Requests: Add Resource-Scoped Constraints

kiosion opened the pull request:
## Context
Add `ResourceAccessID` type for use in Access Request-related contexts, wrapping ResourceIDs with `ResourceConstraints` info, allowing narrower scoping of resource-based Access Requests at creation, and enforcement of requested scope at authorization time.

When `ResourceAccessID`s contain `ResourceConstraints`:
- At request creation: Append new `MatcherFromConstraints` at `roleSet.checkAccess` call to exclude roles incompatible with provided constraints.
- At storage of generated `tlsca.Identity`/`sshca.Identity` certs: Add constrained `ResourceAccessID`s to new cert extension field, and, if only constrained `ResourceAccessID`s are present, add a placeholder/sentinel resourceID to existing `AllowedResourceIDs` field to prevent older Proxies/Auth from interpreting the empty list on the cert as "no resource-specific restrictions".
- At authorization: Wrap existing `RoleMatcher`s to scope principals to provided constraints prior to invoking existing matcher logic.

Initial scope is AWS Console application resources.

## Related material
- [RFD 228](https://github.com/gravitational/teleport/blob/master/rfd/0228-resource-scoped-constraints-in-access-requests.md)
- [RFD 230](https://github.com/gravitational/teleport/blob/master/rfd/0230-component-feature-advertisement.md)
- Related to #58307
- Requires https://github.com/gravitational/teleport/pull/61168
- Required for https://github.com/gravitational/teleport.e/pull/7547
- Required for https://github.com/gravitational/teleport/pull/62152

kiosion committed:
feat: Extend `AWSRole` struct w/ `RequiresRequest`

- Differentiate between already-granted vs. requestable AWSRoles when
  including requestable resources in `clusterUnifiedResourcesGet`
  req

kiosion committed:
feat: Implement Resource Constraints support for Proxy, Auth

smallinsky commented on api/types/access_request.go:
```diff
@@ -414,14 +422,22 @@ func (r *AccessRequestV3) CheckAndSetDefaults() error {
 	}
-	if len(r.GetRoles()) == 0 && len(r.GetRequestedResourceIDs()) == 0 {
+	if r.Spec.RequestedResourceAccessIDs == nil {
+		r.Spec.RequestedResourceAccessIDs = []ResourceAccessID{}
```
Could you explain why ?

kiosion committed:
fixup: Fix tests

bl-nero commented on api/types/resource_ids.go:
```diff
@@ -189,3 +291,29 @@ func ResourceIDsFromString(raw string) ([]ResourceID, error) {
+//
+// This is necessary when handling requests/certs containing only [ResourceAccessID] s,
+// as an empty list of [types.AccessRequest.AllowedResourceIDs] is interpreted as
+// "no resource-specific restrictions".
```
Do I understand correctly that this is for backwards compatibility with an auth server that would not support the new `AllowedResourceIDs` field? So it would be interpreted by the legacy code as allowing only access to a non-existent sentinel resource?

If this is the case, can't we also use the sentinel marker as resource kind so that we don't accidentally allow access to a node called `__SENTINEL__`?

smallinsky replied:
> // CreateSentinelResourceID creates a [ResourceID] that acts as a sentinel valueAdd a comment on  lines R297 to R301Add diff commentMarkdown input:  edit mode selected.WritePreviewAdd a suggestionHeadingBoldItalicQuoteCodeLinkUnordered listNumbered listTask listMentionReferenceSaved repliesAdd FilesPaste, drop, or click to add filesCancelAdd review commentReturn to code
// or placeholder, indicating the absence of a real resource ID.
//
// This is necessary when handling requests/certs containing only [ResourceAccessID] s,
// as an empty list of [types.AccessRequest.AllowedResourceIDs] is interpreted as
// "no resource-specific restrictions".


This is a bit unclear what is the purpose of sentinel value.

I think that the clue of this is to invalidate the ResourceID field. So if we have a new cluster taht support RequestIDWithConstraints and a new Contrainted access request was created then after a cluster downgrade we want to ensure that this new access request will not be handled as access request to all rescues and by setting the ResourceID.Name to a invalid value we are making sure that after cluster is downgraded the request with constraints are not grant any access in legacy flow. 


@kiosion Does this is correct ? 

Could you add a comment here and when the `IsSentinelResourceID` call is used so this flow is be more explicit for the reader ? 

Also: Can we remove the sentinel  in the future in Teleport v20 ? If yes we can add a TODO lablel.

bl-nero commented on api/types/resource_ids.go:
```diff
@@ -83,6 +85,78 @@ func (id *ResourceID) validateK8sSubResource() error {
+	zipped := make([]ResourceAccessID, 0, totalLen)
+	for _, id := range ids {
+		zipped = append(zipped, ResourceAccessID{Id: id})
+	}
```
Consider reusing `ResourceIDsToResourceAccessIDs`. This will protect from modifying one and not another if the semantics of conversion change at any point.

bl-nero commented on api/types/access_request.go:
```diff
@@ -487,14 +503,44 @@ func (r *AccessRequestV3) SetRevision(rev string) {
+	}
+	return wrapped
+}
+
```
Consider adding tests for non-trivial methods.

bl-nero commented on api/types/resource_ids.go:
```diff
@@ -189,3 +291,29 @@ func ResourceIDsFromString(raw string) ([]ResourceID, error) {
+	return id.ClusterName == ResourceIDSentinelValue &&
+		id.Kind == KindNode &&
+		id.Name == ResourceIDSentinelValue
+}
```
Please add tests to new non-trivial public functions.

bl-nero reviewed (commented):
Reviewed roughly half of it, so publishing comments. Will review the rest on Friday.

bl-nero commented on lib/services/access_request.go:
```diff
@@ -2356,19 +2450,31 @@ func (m *RequestValidator) pruneResourceRequestRoles(
+	for _, resource := range underlyingResources {
+		var constraints *types.ResourceConstraints
+		for _, r := range requestedResourceAccessIDs {
+			if rid := r.GetResourceID(); rid.Kind != resource.GetKind() || rid.Name != resource.GetName() {
```
I've seen this RedsourceID to resource matching conditional at least twice, consider extracting it to a separate function.

bl-nero commented on lib/services/split_access_checker.go:
```diff
@@ -73,7 +73,7 @@ type ScopedAccessCheckerSubset interface {
 	RoleNames() []string
 	CertificateFormat() string
-	GetAllowedResourceIDs() []types.ResourceID
+	GetAllowedResourceIDs() []types.ResourceAccessID
```
Consider renaming to `GetAllowedResourceAccessIDs`.

bl-nero commented on lib/tlsca/ca.go:
Just as above, please make sure that there's test coverage for these changes.

kiosion commented on api/types/access_request.go:
```diff
@@ -487,14 +503,44 @@ func (r *AccessRequestV3) SetRevision(rev string) {
-	return append([]ResourceID{}, r.Spec.RequestedResourceIDs...)
+	ids := make([]ResourceID, 0, len(r.Spec.RequestedResourceIDs))
+	for _, id := range r.Spec.RequestedResourceIDs {
+		if !IsSentinelResourceID(id) {
```
The intent was to prevent the front-end ever displaying the sentinel value. I'll verify again this works as intended, though.

kiosion replied:
Sorry, forgot to update – We only want the sentinel value present in certs as a safeguard for versions of Auth prior to these changes; filtering it out if present via this getter should be fine. In any case, when Upserting a req, we call `CheckAndSetDefaults`, which ensures it's in place if `len(RequestedResourceIDs) == 0 && len(RequestedResourceAccessIDs) > 0`. I do agree it feels odd to have `r.Spec.RequestedResourceIDs` potentially yield a different value than the getter, though, but I'm not sure of a cleaner way to approach this without filtering for the sentinel at each site the getter is currently used... WDYT?

kiosion commented on lib/auth/sessions.go:
```diff
@@ -78,7 +78,7 @@ type NewWebSessionRequest struct {
 	AccessRequests []string
 	// RequestedResourceIDs optionally lists requested resources
-	RequestedResourceIDs []types.ResourceID
+	RequestedResourceIDs []types.ResourceAccessID
```
True, will do that instead.

kiosion commented on lib/sshca/identity.go:
Can do.

kiosion committed:
test: Add tests for resourceaccessid-related helpers

kiosion committed:
fixup: Address code review feedback

smallinsky commented on lib/sshca/identity.go:
```diff
@@ -453,7 +469,21 @@ func DecodeIdentity(cert *ssh.Certificate) (*Identity, error) {
-		ident.AllowedResourceIDs = resourceIDs
+		filteredResourceIDs := make([]types.ResourceID, 0, len(resourceIDs))
+		for _, rid := range resourceIDs {
+			if types.IsSentinelResourceID(rid) {
```
Could you add a comment to the reader explaining this behavior.

smallinsky commented on api/types/resource_ids.go:
```diff
@@ -83,6 +85,76 @@ func (id *ResourceID) validateK8sSubResource() error {
+	return nil
+}
+
+func (r *ResourceAccessID) GetResourceID() ResourceID {
```
nit: go docs

smallinsky commented on lib/services/access_checker.go:
```diff
@@ -542,15 +546,32 @@ func (a *accessChecker) CheckAccess(r AccessCheckable, state AccessState, matche
 	return trace.Wrap(a.RoleSet.checkAccess(r, a.info.Traits, state, matchers...))
 }
 
+func (a *accessChecker) matchAllowedRID(r AccessCheckable) (*types.ResourceAccessID, bool) {
```
Seems to be unused

smallinsky commented on lib/services/access_checker.go:
```diff
@@ -453,11 +453,11 @@ func NewAccessCheckerForRemoteCluster(ctx context.Context, localAccessInfo *Acce
 		// certificate does not contain a list of specifically allowed
 		// resources, only role-based access control is used
-		return nil
+		return nil, nil
```
Can we avoid returning nil, nil 

the caller needs to remember to validate the retuned  `types.ResourceAccessID` value if this is not nil  after err != nil. This is considered as antipattern where if the err != nil the value value should be valid.

smallinsky reviewed (commented):
LGTM when remaining comment will be addressed.

kiosion committed:
test: Update identity/events tests for new `ResourceAccessID` fields

kiosion committed:
fixup: Expand/improve godocs/comments for sentinel ResourceID handling

kiosion committed:
fixup: Address code review feedback

kiosion committed:
feat: Conv ResourceID->ResourceAccessID at cert decoding

- Convert ResourceIDs present on certs to ResourceAccessIDs at decoding
  time, rather than at each callsite. Update tests/usages of tlsca/sshca
  Identities to reflect this.
